src/umpire/strategy/DynamicPoolList.cpp

```diff
@@ -178,14 +178,15 @@ PoolCoalesceHeuristic<DynamicPoolList> DynamicPoolList::percent_releasable_hwm(i
   if (percentage == 0) {
     return [=](const DynamicPoolList& UMPIRE_UNUSED_ARG(pool)) { return 0; };
   } else if (percentage == 100) {
-    return
-        [=](const strategy::DynamicPoolList& pool) { return pool.getCurrentSize() == 0 ? pool.getHighWatermark() : 0; };
+    return [=](const strategy::DynamicPoolList& pool) {
+      return pool.getCurrentSize() == 0 ? pool.getActualHighwaterMark() : 0;
+    };
   } else {
     float f = (float)((float)percentage / (float)100.0);
     return [=](const strategy::DynamicPoolList& pool) {
       // Calculate threshold in bytes from the percentage
       const std::size_t threshold = static_cast<std::size_t>(f * pool.getActualSize());
-      return pool.getReleasableSize() >= threshold ? pool.getHighWatermark() : 0;
+      return pool.getReleasableSize() > threshold ? pool.getActualHighwaterMark() : 0;
```

Comment:
Here and in `QuickPool.cpp`, I had to change the comparison to `>` from `>=` to avoid infinite recursion. With `getActualHighwaterMark` https://github.com/LLNL/Umpire/blob/bf2df8d46f6e8918b7402a45b0f95127c3121478/tests/unit/strategy/pool_heuristics_tests.cpp#L116 ends up exhausting the stack when coalescing because `getReleasableSize() == threshold` after coalescing. Since the last operation of coalescing is deallocating (https://github.com/LLNL/Umpire/blob/bf2df8d46f6e8918b7402a45b0f95127c3121478/src/umpire/strategy/QuickPool.cpp#L335), which then checks for coalescing again, you end up with infinite recursion.

If the original behaviour should be kept then I suppose allocation and creation of new blocks could be separated, such that `do_coalesce` could release memory, allocate a new block and return, instead of going via `allocate` and `deallocate` to create a new block.

What do you think is best here?

Reply:
Perhaps https://github.com/LLNL/Umpire/blob/bf2df8d46f6e8918b7402a45b0f95127c3121478/src/umpire/strategy/QuickPool.cpp#L325 should be `getReleasableBlocks() > 1` instead? If there's only one or zero blocks there's no coalescing that can be done.

Reply:
I've reverted the `>=` to `>` change with the latest additions. It's no longer an issue when using `getAlignedHighwaterMark` instead of `getActualHighwaterMark`.

The second comment above is still relevant, but is not a must for this PR.